app/com/arpnetworking/metrics/portal/scheduling/impl/UnboundedPeriodicSchedule.java

```diff
@@ -0,0 +1,171 @@
+/*
+ * Copyright 2020 Dropbox, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.arpnetworking.metrics.portal.scheduling.impl;
+
+import com.arpnetworking.commons.builder.OvalBuilder;
+import com.arpnetworking.logback.annotations.Loggable;
+import com.arpnetworking.metrics.portal.scheduling.Schedule;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.MoreObjects;
+import com.google.common.base.Objects;
+import net.sf.oval.constraint.Min;
+import net.sf.oval.constraint.NotNull;
+
+import java.time.Clock;
+import java.time.Duration;
+import java.time.Instant;
+import java.time.temporal.ChronoUnit;
+import java.util.Optional;
+
+/**
+ * A schedule for a job that repeats periodically, without any bounds or regard
+ * for the last completed time.
+ * <p>
+ * The next run for a unbounded periodic schedule is always the start of the
+ * most recent period of the schedule, as aligned with the start of the epoch.
+ * <p>
+ * <b>WARNING:</b>
+ * This behavior means that jobs with this schedule will ignore previously missed runs
+ * - if you need backfills you should instead use a standard {@link PeriodicSchedule}.
+ *
+ * @author Christian Briones (cbriones at dropbox dot com)
+ */
+@Loggable
+public final class UnboundedPeriodicSchedule implements Schedule {
+
+    private final ChronoUnit _period;
+    private final long _periodCount;
+    private final Clock _clock;
+
+    private UnboundedPeriodicSchedule(final Builder builder) {
+        _period = builder._period;
+        _periodCount = builder._periodCount;
+        _clock = builder._clock;
+    }
+
+    @Override
+    public Optional<Instant> nextRun(final Optional<Instant> lastRun) {
+        final Duration fullPeriod = Duration.of(_periodCount, _period);
+
+        final Instant now = _clock.instant();
+        final Instant start = lastRun
+                .filter(run -> run.compareTo(now) >= 0)
```

Comment:
This eliminates any runs in the past, right? -- this is to protect us agains the system clock moving. e.g. backwards and repeating work units or forwards and trying to catch-up (vs skip).

My only concern on the backwards is that it would effectively create a clock-time gap when nothing is run. Effectively, I am wondering whether the difference between report and alert scheduling is cpu time vs clock time. Alerts are evaluated every 30 seconds, vs reports are delivered hourly/daily based on the clock.

I'm not 100% convinced this is the difference, although I suspect it is the source of some of these issues.

Reply:
Correct - it's to protect against the backwards system clock and also the case where the last run was already the most recent period so we can ignore it as completed.

That might be the case. The other thing I'm noticing is that there are two distinct types of schedules - ones like this where all you care about is the job running every period but you don't really care about history, and the other case where you want to ensure that every missed period up to the current time gets evaluated (e.g. reporting backfill).